components/brave_shields/browser/brave_shields_util.cc

```diff
@@ -332,25 +332,69 @@ void SetFingerprintingControlType(Profile* profile,
   if (!primary_pattern.IsValid())
     return;
 
-  auto* map = HostContentSettingsMapFactory::GetForProfile(profile);
-  map->SetContentSettingCustomScope(
-      primary_pattern, ContentSettingsPattern::Wildcard(),
-      ContentSettingsType::PLUGINS, kFingerprintingV2,
-      GetDefaultAllowFromControlType(type));
+  // Cleanup previous state.
+  HostContentSettingsMapFactory::GetForProfile(profile)
+      ->SetContentSettingCustomScope(
+          primary_pattern,
+          ContentSettingsPattern::Wildcard(),
+          ContentSettingsType::PLUGINS,
+          kFingerprintingV2,
+          CONTENT_SETTING_DEFAULT);
+  HostContentSettingsMapFactory::GetForProfile(profile)
+      ->SetContentSettingCustomScope(
+          primary_pattern,
+          ContentSettingsPattern::FromString("https://balanced/*"),
+          ContentSettingsType::PLUGINS,
+          kFingerprintingV2,
+          CONTENT_SETTING_DEFAULT);
+
+  // Secondary pattern is only set for balanced mode.
+  ContentSetting content_setting = GetDefaultAllowFromControlType(type);
+  auto secondary_pattern = ContentSettingsPattern::Wildcard();
+  if (content_setting == CONTENT_SETTING_DEFAULT) {
+    content_setting = CONTENT_SETTING_BLOCK;
+    secondary_pattern =
+        ContentSettingsPattern::FromString("https://balanced/*");
+  }
+
+  HostContentSettingsMapFactory::GetForProfile(profile)
+      ->SetContentSettingCustomScope(
+          primary_pattern,
+          secondary_pattern,
+          ContentSettingsType::PLUGINS,
+          kFingerprintingV2,
+          content_setting);
 
   RecordShieldsSettingChanged();
 }
 
 ControlType GetFingerprintingControlType(Profile* profile, const GURL& url) {
   auto* map = HostContentSettingsMapFactory::GetForProfile(profile);
   ContentSetting setting = map->GetContentSetting(
-        url, GURL(), ContentSettingsType::PLUGINS, kFingerprintingV2);
-  if (setting == CONTENT_SETTING_BLOCK) {
-    return ControlType::BLOCK;
-  } else if (setting == CONTENT_SETTING_ALLOW) {
+      url, GURL(), ContentSettingsType::PLUGINS, kFingerprintingV2);
+  ContentSetting balanced_setting =
+      map->GetContentSetting(url, GURL("https://balanced/"),
+                             ContentSettingsType::PLUGINS, kFingerprintingV2);
+
+  if (balanced_setting == CONTENT_SETTING_ALLOW) {
     return ControlType::ALLOW;
   }
-  return ControlType::DEFAULT;
+
+  // Clean state.
+  if (balanced_setting == CONTENT_SETTING_DEFAULT) {
+    return ControlType::DEFAULT;
+  }
+
+  if (balanced_setting == CONTENT_SETTING_BLOCK) {
+    if (setting == CONTENT_SETTING_BLOCK) {
+      // There is no way to determine whether site setting is balanced or
+      // aggressive(block) here.
```

Comment:
I don't know how to determine whether site setting is balanced or aggressive in here.